### DISCUSSION TITLE
nx.from_numpy_array() with non-square arrays

### DISCUSSION THREAD

**Discussion by ns-rse** [networkx/networkx#7093] [Q&A]
Hi,

I'm looking to use `networkx` in conjunction with [Skan](https://github.com/jni/skan) and would like to seamlessly convert from the internal `Skeleton` object to `Networkx` object and back.

With Skan the `Skeleton` class can takes arrays as input and stores them as a property which can be accessed [`Skeleton.source_image`](https://skeleton-analysis.org/stable/api/skan.csr.html#skan.csr.Skeleton.source_image).

I figured I could therefore leverage [`Networkx.from_numpy_array()`](https://networkx.org/documentation/stable/reference/generated/networkx.convert_matrix.from_numpy_array.html) and [`Networkx.to_numpy_array()`](https://networkx.org/documentation/stable/reference/generated/networkx.convert_matrix.to_numpy_array.html) to do this conversion back and forth.

This works ok on a simple test array...

```python
nested_list = [[0, 1],
               [1, 0]]
square_array = np.array(nested_list)
# Make a square skeleton
square_skel = Skeleton(square_array)
nested_list = [[0, 1],
               [1, 0]]
square_array = np.array(nested_list)

# Make a square skeleton
square_skel = Skeleton(square_array)
square_nx_from_array = nx.from_numpy_array(square_skel.skeleton_image)
nx.draw(square_nx_from_array, pos=nx.circular_layout(square_nx_from_array))
```

![image](https://github.com/networkx/networkx/assets/17566406/7c1d23f3-c6b5-4a63-9b4c-2a0813f00862)

However, if the input array isn't square (e.g. rectangular) it raises a `NetworkXError: Adjacency matrix not square: nx,ny=(2, 3)`.

```python
nested_list = [[0, 1, 0],
               [1, 0, 1]]
rectangle_array = np.array(nested_list)

# Make a rectangle skeleton
rectangle_skel = Skeleton(rectangle_array)
# Make a rectangle NetworkX
rectangle_nx_from_array = nx.from_numpy_array(rectangle_skel.skeleton_image)
nx.draw(rectangle_nx_from_array, pos=nx.circular_layout(square_nx_from_array))


Truncated Traceback (Use C-c C-$ to view full TB):
File ~/.virtualenvs/skan-pyproject/lib/python3.11/site-packages/networkx/convert_matrix.py:1104, in from_numpy_array(A, parallel_edges, create_using)
   1102 n, m = A.shape
   1103 if n != m:
-> 1104     raise nx.NetworkXError(f"Adjacency matrix not square: nx,ny={A.shape}")
   1105 dt = A.dtype
   1106 try:

NetworkXError: Adjacency matrix not square: nx,ny=(2, 3)
```

My initial thought was to pad the non-square array to make it square...

```python
def pad_non_square_array(adjacency_array: npt.NDArray) -> npt.NDArray:
    """Pad a non-square adjacency array with zeros so that it is square.

    Parameters
    ----------
    adjacency_array : npt.NDArray
        2-D Numpy adjaceny array with uneven dimensions.

    Returns
    -------
    npt.NDArray
        Padded array where ncols equals nrows.
    """
    try:
        nrows, ncols = adjacency_array.shape
    except ValueError:
        nrows, ncols = (1, len(adjacency_array))
    if nrows / ncols < 1:
        pad_width = ((0, ncols - nrows), (0, 0))
    elif nrows / ncols > 1:
        pad_width = ((0, 0), (0, nrows - ncols))
    else:
        pad_width = ((0, 0), (0, 0))
    return np.pad(adjacency_array, pad_width)
padded_array = pad_non_square_array(rectangle_skel.skeleton_image)
print(f"Padded Array :\n{padded_array}")
padded_nx_from_array = nx.from_numpy_array(padded_array)
nx.draw(padded_nx_from_array)
Padded Array :
[[0 1 0]
 [1 0 1]
 [0 0 0]]
```

![image](https://github.com/networkx/networkx/assets/17566406/f45d2d61-de06-470d-9d1e-8db7c53fb0ac)

This is looking promising so I tried it on a larger, but relatively simple, skeleton from out test suite...

```python
# Define a quick plotting function
draw_kwargs = {"with_labels": True}
def networkx_plot(adjacency: np.ndarray, draw_kwargs=draw_kwargs):
    """Pad an adjancency array, convert to Skeleton then NetworkX graph and plot."""
    padded = pad_non_square_array(adjacency)
    skeleton = Skeleton(padded)
    print(f"Padded Array :\n{skeleton.skeleton_image}")
    skeleton_nx = nx.from_numpy_array(skeleton.skeleton_image)
    nx.draw(skeleton_nx, pos=nx.circular_layout(skeleton_nx), **draw_kwargs)

Padded Array :
[[False, False, False,  True, False, False, False],
 [False, False, False,  True, False, False, False],
 [False, False, False,  True, False, False, False],
 [ True,  True,  True,  True,  True,  True,  True],
 [False, False, False, False, False, False, False],
 [False, False, False, False, False, False, False],
 [False, False, False, False, False, False, False]])
```

![image](https://github.com/networkx/networkx/assets/17566406/88252ca0-4fb0-4d35-880b-9553bee1226d)

This looks a little strange as I was expecting only four nodes. @jni put together the following function to construct `Networkx.Graph` from a `Skeleton`...

```python
def skeleton_to_nx_orig(
        skeleton: Skeleton, summary: pd.DataFrame | None = None
        ) -> nx.Graph:
    """Convert a Skeleton object to a networkx Graph.

    Parameters
    ----------
    skeleton : Skeleton
        Skeleton object to be converted.
    summary : pd.DataFrame, optional
        If the skeleton has already been summarized() pass it in, if not provided it will be summarized.

    Raises
    ------
    TypeError
        If a Numpy array is passed by mistake an error is raised.

    Returns
    -------
    nx.Graph
        Returns a networkx Graph
    """
    if isinstance(skeleton, np.ndarray):
        raise TypeError(
                'You have passed a Numpy array, please convert to Skeleton first.'
                )
    if summary is None:
        summary = summarize(skeleton)
    summary_future = summary.rename(columns=lambda s: s.replace('-', '_'))
    g = nx.Graph()
    for row in summary_future.itertuples(name='Edge'):
        i, j = row.node_id_src, row.node_id_dst
        g.add_edge(i, j, **row._asdict())
        g.edges[i, j]['path'] = skeleton.path_coordinates(row.Index)
        g.nodes[i]['pos'] = skeleton.coordinates[i]
        g.nodes[j]['pos'] = skeleton.coordinates[j]
    return g
```

If I use this to construct the graph the resulting plot is more what I'd expect (although I'm still very naive about Network Graphs and associated theory and can't get my head round why the output `nx.to_numpy_array()` adjacency matrix is as shown).

```python
skeleton0_skel = Skeleton(skeleton0)
skeleton0_nx = skeleton_to_nx_orig(skeleton0_skel)
print(skeleton0_nx)
print(f"NetworkX to numpy array :\n{nx.to_numpy_array(skeleton0_nx)}")
nx.draw(skeleton0_nx, pos=nx.circular_layout(skeleton0_nx), **draw_kwargs)

Graph with 4 nodes and 3 edges
NetworkX to numpy array :
[[0. 1. 0. 0.]
 [1. 0. 1. 1.]
 [0. 1. 0. 0.]
 [0. 1. 0. 0.]]
```
![image](https://github.com/networkx/networkx/assets/17566406/91a427ff-b13b-4fe6-a86f-42f7aa527194)

### Why does `nx.from_numpy_array()` raise an error with non-square arrays?

Is there a reason that the Numpy adjacency array should be square? If so, what it the best way to avoid the artifacts this seems to introduce?

### Why does padding of an adjacency graph affect the NetworkX Graph?

Its unclear to me why padding with extra empty rows increases the number of nodes and edges in the graph.

I found an [old question](https://github.com/networkx/networkx/discussions/4692) that seems to pre-date the introduction of `nx.from_numpy_array()` that mentions directions can influence things, but don't know enough to be able to say whether that is related.

**Comment by MridulS** [comment]
> Why does nx.from_numpy_array() raise an error with non-square arrays?

`from_numpy_array` is used to create an adjacency matrix for the graph. An [adjacency matrix](https://en.wikipedia.org/wiki/Adjacency_matrix) is by definition a square matrix.  

> Why does padding of an adjacency graph affect the NetworkX Graph?

The padding is adding more nodes and edges! It's changing the inherent structure of the `source_image`.

From a *very* cursory reading for Skan documentation, you can probably use the [skan.csr.Skeleton.graph](https://skeleton-analysis.org/stable/api/skan.csr.html#skan.csr.Skeleton.graph) attribute to read the csr graph using [from_scipy_sparse_array](https://networkx.org/documentation/stable/reference/generated/networkx.convert_matrix.from_scipy_sparse_array.html) into networkx directly.

**Reply by ns-rse** [reply]
Brilliant, thank you for taking the time to explaining and linking to more information, I'll investigate your suggestion.

Really appreciate this, thanks @MridulS :+1:

**Comment by jni** [comment]
You're correct in your interpretation @MridulS. @ns-rse pretty much the whole reason skan exists is because networkx can't build pixel graphs from images — the `from_numpy_array` method in nx expects a completely different kind of array: an adjacency matrix, as @MridulS points out. This confusion appears to stem from the fact that NumPy arrays can represent many kinds of data — in skan, an image, in networkx, an adjacency matrix, in scikit-learn, an array of samples x features... These are not interchangeable.

**Reply by ns-rse** [reply]
Its completely my misunderstanding and lack of domain knowledge.
